### PR TITLE
feat(config): read config from branches in baseBranches

### LINKED ISSUE #2496: Read configuration from the branch(es) in `baseBranches`
I'm sorry in advance if this is a duplicate. I tried searching for some keywords but nothing came up.

### **What would you like Renovate to be able to do?**
Read `renovate.json` configuration from the branch(es) in `baseBranches` instead of the repository's default branch, so changes in `baseBranches` are applied instantaneously.

At [`pymedusa/Medusa`](https://github.com/pymedusa/Medusa) we have two "main" branches: `master` and `develop`.
`master` is the repo's default branch
`develop` is used for development **and** is the single branch listed under [`baseBranches` on the `renovate.json`](https://github.com/pymedusa/Medusa/blob/develop/renovate.json#L5-L7) config file.
(We use the GitHub app if that matters)

When we change the renovate configuration we **have** to wait until the next release to `master` in order for the changes to be applied.

### **Describe the solution you'd like**
Would it be possible, to have a config option to "tell" the renovate bot to read the configuration (sans-`baseBranches` and other "risky" options?) from the `baseBranch` that is configured on the default branch?
So when it is enabled, the bot will consider the configuration defined in `renovate.json` on the `baseBranch` it is creating the dependency update branch and PR.

So for example,
if the default branch's `renovate.json` currently doesn't contain a `labels` property,
but the `baseBranch`'s `renovate.json` contains `"labels": ["update dep"]` - it should apply the label to the PR being opened against that `baseBranch`.

### **Describe alternatives you've considered**
Switching the default branch to `develop`, which is not viable in our case.

### **Additional context**
N/A

-----

We've been using the renovate GitHub app in our Python+JS project for two months now and we're very pleased! Thank you for this amazing service and support!
/cc @OmgImAlexis

### PR DESCRIPTION
**Pull request by olegkrivtsov**
## Changes:

This is work in progress for Read configuration from the branch(es) in baseBranches #2496

I decided to push and request for some review/advice. I'd like to confirm that I'm moving in the right direction, since I've modified the interface for `getJsonFile()` and had to modify many function in `lib/platform`. Could you please let me know if that's what you expect me to do?

## Context:

Closes #2496 

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/platform/bitbucket-server/index.ts
Comment (viceice) [inline lib/platform/bitbucket-server/index.ts:127]:
```suggestion
  branch?: string
```

File: lib/platform/bitbucket-server/index.ts
Comment (viceice) [inline lib/platform/bitbucket-server/index.ts:144]:
```suggestion
  branch?: string
```

File: lib/platform/bitbucket/index.ts
Comment (viceice) [inline lib/platform/bitbucket/index.ts:114]:
```suggestion
  branch?: string
```

File: lib/platform/bitbucket/index.ts
Comment (viceice) [inline lib/platform/bitbucket/index.ts:126]:
```suggestion
  branch?: string
```

File: lib/platform/gitlab/index.ts
Comment (viceice) [inline lib/platform/gitlab/index.ts:161]:
```suggestion
  branch?: string
```

File: lib/platform/gitea/index.ts
Comment (viceice) [inline lib/platform/gitea/index.ts:224]:
```suggestion
    branch?: string
```

File: lib/platform/gitea/index.ts
Comment (viceice) [inline lib/platform/gitea/index.ts:215]:
```suggestion
    branch?: string
```

File: lib/platform/azure/index.ts
Comment (viceice) [inline lib/platform/azure/index.ts:138]:
```suggestion
  branch?: string
```

File: lib/platform/azure/index.ts
Comment (viceice) [inline lib/platform/azure/index.ts:117]:
```suggestion
  branch?: string
```

Should match interface declaration

**Review by viceice — CHANGES_REQUESTED** [review]
I think you should stop on this feature, as #11565 will conflict and is nearly ready to merge

**Comment by viceice** [comment]
> I think you should stop on this feature, as #11565 will conflict and is nearly ready to merge

Nope, no conflict, you can continue. Maybe we should name the new parameter `branchOrTag` as it should also support tag refs

**Comment by olegkrivtsov** [comment]
Hi @rarkins I'm currently working on unit tests for `getJsonFile()` on supported platforms and it seems that Bitbucket Server API and Bitbucket Cloud API don't support getting file content from a branch. At least I see this from this documentation:

https://developer.atlassian.com/cloud/bitbucket/rest/api-group-source/#api-repositories-workspace-repo-slug-src-commit-path-get

I can only get file contents from the master branch.

For Bitbucket Server API I couldn't find a good source of information. I just found this one and it wasn't good enough: https://docs.atlassian.com/bitbucket-server/rest/7.17.0/bitbucket-git-rest.html

For now I decided to throw an Error if someone specifies the `branchOrTag` argument for `getJsonFile()` on those platforms. Is that a good solution?

**Comment by viceice** [comment]
Can't we simply replace `HEAD` for bb cloud?
https://github.com/renovatebot/renovate/blob/db6bc92d1bc55198e9688e24d87baa2b45f610ca/lib/platform/bitbucket/index.ts#L111-L120

**Comment by viceice** [comment]
bb-server:
https://docs.atlassian.com/bitbucket-server/rest/5.3.0/bitbucket-rest.html#idm46551990789376

Simply add `at` parameter

https://github.com/renovatebot/renovate/blob/db6bc92d1bc55198e9688e24d87baa2b45f610ca/lib/platform/bitbucket-server/index.ts#L124-L138

**Review by rarkins — COMMENTED** [review]
These suggestions got stuck on "Pending" previously, adding them now in case they are still relevant

File: lib/platform/azure/index.ts
Comment (olegkrivtsov) [reply lib/platform/azure/index.ts]: Hi @rarkins I renamed `branch` into `branchOrTag` as suggested by @viceice. Do you request renaming from `branchOrTag` into `branchName`?

Comment (rarkins) [reply lib/platform/azure/index.ts]: Fine to use branchOrTag

File: lib/platform/bitbucket-server/index.ts
Comment (olegkrivtsov) [reply lib/platform/bitbucket-server/index.ts:126]: I'll try to do that

Comment (olegkrivtsov) [reply lib/platform/bitbucket-server/index.ts:126]: @rarkins do you want this refactoring only for `getRawFile` or for `getJsonFile` as well?

Comment (rarkins) [reply lib/platform/bitbucket-server/index.ts:126]: Yes, but maybe separate the refactoring into a separate (initial) PR. We should have consistent handling of `null`/`undefined` repository values across platforms.

File: lib/workers/repository/process/index.ts
Comment (olegkrivtsov) [reply lib/workers/repository/process/index.ts:34]: Sorry, it is not clear how to do that. Maybe you can give an example?

File: lib/platform/bitbucket-server/index.ts
Comment (olegkrivtsov) [reply lib/platform/bitbucket-server/index.ts:126]: @rarkins I'm going to work on this refactoring in a separate PR and push it once ready.

File: lib/workers/repository/process/index.ts
Comment (olegkrivtsov) [reply lib/workers/repository/process/index.ts:34]: @rarkins I agree merge is needed, because some base config fields are missing and it doesn't work on a real repo. I'm trying to fix.

**Review by viceice — CHANGES_REQUESTED** [review]
I would suggest to split this pr into smaller parts.

1. one pr to add the optional `branchOrTag`, which is unused at platforms, unify null handling
2. one pr per platform to add the `branchOrTag` parameter (so we can easier test)
3. this pr to support the `baseBranches` feature only

**Comment by olegkrivtsov** [comment]
@viceice I'll do as you suggested.

File: lib/workers/repository/process/index.ts
Comment (olegkrivtsov) [reply lib/workers/repository/process/index.ts:34]: I'd like to ask for an advice from you. I was thinking how to merge the global config we have from the global worker and the branch config we get from `getJsonFile()` AND *remove* the master branch config we already parsed on the repo init stage. I see a possible solution: modify this function:

https://github.com/renovatebot/renovate/blob/51fa664ba42f75bab334d3de74d61bd612aac0ca/lib/workers/repository/init/merge.ts#L167

and add the second optional argument `branchName`. This function does exactly what I need - it merges the global config with the `renovate.json` file it reads from the default branch. Instead it will detect/read config file from the branch I'll specify. 

Then I'll need to call `mergeRenovateConfig()` and it will do the merge I need.

I'd like to confirm with you first before continue implementing this solution, because it may have side effects I can't see for now.

Comment (olegkrivtsov) [reply lib/workers/repository/process/index.ts:38]: Hi @rarkins could you please advice on this? I'd like to modify `mergerRenovateConfig()` since it does a lot of things with the config: if the config has `extends`, merges the extended parts, and others. I would proceed with this solution if you approve.

Comment (olegkrivtsov) [reply lib/workers/repository/process/index.ts:46]: I did my best try to merge the global config with the config read from branch. It complaints something about `fileMatch is not iterable` when I run it on the real repo. If you could review my latest changes and advice where I should improve the code, that would help me a lot. I'm a bit stuck with this one.

File: lib/workers/repository/process/index.spec.ts
Comment (viceice) [inline lib/workers/repository/process/index.spec.ts:2]:
```suggestion
import { platform, RenovateConfig, getConfig, git, mocked } from '../../../../test/util';
```

File: lib/workers/repository/process/index.spec.ts
Comment (viceice) [inline lib/workers/repository/process/index.spec.ts:39]:
```suggestion
      platform.getJsonFile.mockResolvedValueOnce({});
```

Platform is already mocked

File: lib/workers/repository/process/index.spec.ts
Comment (viceice) [inline lib/workers/repository/process/index.spec.ts:43]: Use `toEqual` instead

File: lib/workers/repository/process/index.ts
Comment (viceice) [inline lib/workers/repository/process/index.ts:2]:
```suggestion
import type { GlobalConfig } from '../../../config/global';
```

File: lib/workers/repository/process/index.ts
Comment (viceice) [inline lib/workers/repository/process/index.ts:41]: I don't think you want to merge the global config here, as those values shouldn't be on branch config. 🤔 
```suggestion
    baseBranchConfig = mergeChildConfig(config, baseBranchConfig);
```

File: lib/workers/repository/process/index.ts
Comment (viceice) [reply lib/workers/repository/process/index.ts:45]: Maybe you've some invalid config? can you share the configs? And a stacktrace.

File: lib/workers/repository/process/index.ts
Comment (viceice) [inline lib/workers/repository/process/index.ts:36]: We need a catch block here, if there is a config file name mismatch between branches. Maybe we need to cache it per branch

File: lib/workers/repository/process/index.ts
Comment (rarkins) [reply lib/workers/repository/process/index.ts:45]: I saw that error when a regex manager had no `fileMatch` field

Comment (olegkrivtsov) [reply lib/workers/repository/process/index.ts:52]: After your help it started working on my real repo! Thanks a lot 👍

**Comment by olegkrivtsov** [comment]
Tested on a real repo https://github.com/olegkrivtsov/renovate-test-repo with two base branches: `master` and `dev`. For `dev` it used a different branch prefix as per my branch config.

File: lib/workers/repository/process/index.ts
Comment (viceice) [inline lib/workers/repository/process/index.ts:45]: i think we schould throw an config error here

@rarkins WDYT?

File: lib/workers/repository/process/index.ts
Comment (viceice) [inline lib/workers/repository/process/index.ts:78]: See comment above, so this can reverted

File: lib/workers/repository/process/index.ts
Comment (viceice) [inline lib/workers/repository/process/index.ts:93]: Same here

**Review by viceice — DISMISSED** [review]

File: lib/workers/repository/process/index.ts
Comment (viceice) [inline lib/workers/repository/process/index.ts:17]:
```suggestion
): Promise<RenovateConfig> {
```

File: lib/workers/repository/process/index.ts
Comment (viceice) [inline lib/workers/repository/process/index.ts:33]:
```suggestion
      cache.configFileName;
```
I think the cache should already be populated while reading config from default branch.

File: lib/workers/repository/process/index.ts
Comment (viceice) [inline lib/workers/repository/process/index.ts:27]:
```suggestion
      { baseBranch },
      `Using config from base branch because useBaseBranchConfig option specified`
```

File: lib/workers/repository/process/index.ts
Comment (viceice) [inline lib/workers/repository/process/index.ts:43]:
```suggestion
        { configFileName, baseBranch },
        `Error fetching config file from base branch - possible config name mismatch between branches?`
```

File: lib/workers/repository/process/index.ts
Comment (viceice) [inline lib/workers/repository/process/index.ts:47]: Please throw like this:
https://github.com/renovatebot/renovate/blob/91842073ef58bf51ff9f492c8fce24c81f09fb23/lib/config/secrets.ts#L76-L82

**Review by viceice — DISMISSED** [review]

File: lib/platform/bitbucket-server/index.spec.ts
Comment (rarkins) [inline lib/platform/bitbucket-server/index.spec.ts:2186]:
```suggestion
```

File: lib/platform/bitbucket-server/index.spec.ts
Comment (rarkins) [inline lib/platform/bitbucket-server/index.spec.ts:2200]:
```suggestion
```

File: lib/workers/repository/process/index.ts
Comment (rarkins) [inline lib/workers/repository/process/index.ts:34]: Does this mean that the config file name in the baseBranch must match the file name in the default branch? If so then we should document this.

File: lib/workers/repository/process/index.ts
Comment (rarkins) [inline lib/workers/repository/process/index.ts:60]: Are you sure about this? In theory `baseBranches` should be completely ignored in all but the default branch

**Review by rarkins — CHANGES_REQUESTED** [review]
I think this behavior actually maps to the described "merge" functionality in https://github.com/renovatebot/renovate/issues/2496#issuecomment-419787538 and not "replace". This is because we are merging the base branch's config on top of the default branch's config instead of completely replacing it. If so then we should rename this option.

File: docs/usage/configuration-options.md
Comment (rarkins) [inline docs/usage/configuration-options.md:2586]:
```suggestion
You can set `useBaseBranchConfig` to 'merge' to instruct Renovate to merge the config from each base branch over the top of the config in the default branch.
The config file name in the base branch must be the same as in the default branch.
```

File: lib/config/options/index.ts
Comment (rarkins) [inline lib/config/options/index.ts:702]:
```suggestion
    allowedValues: ['none', 'merge'],
```

We should change this to merge everywhere

File: lib/workers/repository/process/index.ts
Comment (rarkins) [inline lib/workers/repository/process/index.ts:60]:
```suggestion
  if (config.baseBranches.length > 1) {
```

**Review by rarkins — CHANGES_REQUESTED** [review]

File: docs/usage/configuration-options.md
Comment (rarkins) [inline docs/usage/configuration-options.md:2586]:
```suggestion
You can configure `useBaseBranchConfig=merge` to instruct Renovate to merge the config from each base branch over the top of the config in the default branch.
```

File: lib/config/options/index.ts
Comment (rarkins) [inline lib/config/options/index.ts:702]:
```suggestion
    allowedValues: ['merge', 'none'],
```

File: lib/config/options/index.ts
Comment (rarkins) [inline lib/config/options/index.ts:700]:
```suggestion
      'Whether to read configuration from baseBranches instead of only the default branch',
```

File: lib/workers/repository/process/index.ts
Comment (rarkins) [inline lib/workers/repository/process/index.ts:24]: This should be type to catch this type of error. Valid values are `'merge' | 'none'`

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/workers/repository/process/index.ts
Comment (olegkrivtsov) [reply lib/workers/repository/process/index.ts]: I thought that maybe branches would overlap, but now I think it should work without this block

File: lib/workers/repository/process/index.ts
Comment (rarkins) [inline lib/workers/repository/process/index.ts:29]:
```suggestion
      `Merging config from base branch because useBaseBranchConfig=merge`
```

File: docs/usage/configuration-options.md
Comment (rarkins) [inline docs/usage/configuration-options.md:2587]:
```suggestion

The config file name in the base branch must be the same as in the default branch and cannot be `package.json`.
```

File: lib/workers/repository/process/index.ts
Comment (rarkins) [inline lib/workers/repository/process/index.ts:56]:
```suggestion
    // baseBranches value should be based off the default branch
    baseBranchConfig.baseBranches = config.baseBranches;
```

**Review by rarkins — CHANGES_REQUESTED** [review]

**Comment by rarkins** [comment]
@olegkrivtsov do you have a working demo of this functionality?

**Comment by olegkrivtsov** [comment]
@rarkins please check this repo: https://github.com/olegkrivtsov/renovate-test-repo

I've just ran Renovate on it and seems to work as expected (for master and dev branches):

```
$ yarn start
yarn run v1.22.17
$ run-s generate:*
$ node tools/generate-imports.mjs
generating imports
> data/node-js-schedule.json
> node_modules/emojibase-data/en/shortcodes/github.json
$ node -r ts-node/register/transpile-only -- lib/renovate.ts
 INFO: Repository started (repository=olegkrivtsov/renovate-test-repo)
       "renovateVersion": "0.0.0-semantic-release"
 INFO: Dependency extraction complete (repository=olegkrivtsov/renovate-test-repo)
       "baseBranch": "master",
       "stats": {
         "managers": {
           "docker-compose": {"fileCount": 1, "depCount": 1},
           "dockerfile": {"fileCount": 1, "depCount": 1}
         },
         "total": {"fileCount": 2, "depCount": 2}
       }
 INFO: Dependency extraction complete (repository=olegkrivtsov/renovate-test-repo)
       "baseBranch": "dev",
       "stats": {
         "managers": {
           "docker-compose": {"fileCount": 1, "depCount": 1},
           "dockerfile": {"fileCount": 1, "depCount": 1}
         },
         "total": {"fileCount": 2, "depCount": 2}
       }
 INFO: Branch created (repository=olegkrivtsov/renovate-test-repo, branch=renovate/httpd-2.x)
       "commitSha": "6dddbc2"
 INFO: PR created (repository=olegkrivtsov/renovate-test-repo, branch=renovate/httpd-2.x)
       "pr": 29,
       "prTitle": "Update httpd Docker tag to v2"
 INFO: Branch created (repository=olegkrivtsov/renovate-test-repo, branch=devbranch/httpd-2.x)
       "commitSha": "c48afb7"
 INFO: PR created (repository=olegkrivtsov/renovate-test-repo, branch=devbranch/httpd-2.x)
       "pr": 30,
       "prTitle": "Update httpd Docker tag to v2"
 INFO: Repository finished (repository=olegkrivtsov/renovate-test-repo)
       "durationMs": 23369
Done in 50.02s.
```

**Review by rarkins — APPROVED** [review]

**Review by viceice — COMMENTED** [review]
Waiting for docs review from @HonkingGoose , otherwise LGTM

**Comment by HonkingGoose** [comment]
> Waiting for docs review from @HonkingGoose , otherwise LGTM

We're too late, it already got merged. But I saw that @rarkins already suggested a lot of good changes to the docs. 😉

**Pull request merged**
